### PR TITLE
2020-06-16: Add post security-audits

### PR DESCRIPTION
**Pull request by bdemers**
This blog post has the following features:

- [ ] A GitHub Repository with a polished README
- [ ] A title that's approved by Dev Marketing
- [ ] A URL that's approved by Dev Marketing
- [ ] Rendered locally and confirmed that no Markdown typos exist

### CODE REVIEW

File: _source/_posts/2020-06-16-security-audits.adoc [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+In-depth code analysis is great, but when it comes from outside your team, they are generally slow and too costly (time and money) to do against every commit. This often leaves teams to do them "as needed." The catch 22 is—if you are not a security expert—how do you know when they are needed? One option is to schedule them periodically—quarterly, for example—or some other frequency that matches the pace of your development.
+
+== Outside Audit
+
+For the first release of https://github.com/paseto-toolkit/jpaseto[JPaseto] we hired https://paragonie.com/[Paragon Initiative Enterprises] to do a security audit. They had domain expertise, as they helped author the https://paseto.io/rfc/[PASETO RFC]. The turn around was fast, and they came back with a few issues:
+
+. Constant Time Comparison
++
+This was the most critical of the issues they found. The library was making a comparison of a hashed value using Java's `Arrays.equals()` method which fails fast, because it's optimized for speed. The code should have been using a time constant comparison such as `MessageDigest.isEqual()` to prevent timing attacks.
```
Comment (mraible) [inline _source/_posts/2020-06-16-security-audits.adoc:38]: I just learned about the + line in AsciiDoc a few weeks ago! It's pretty slick and can be used to add code blocks into lists too.

File: _source/_posts/2020-06-16-security-audits.adoc [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+. Variable Names
++
+Typically, you don't think of variable names when talking about security issues, but an incorrectly named variable can lead to maintenance issues or improper usage in the future. In this particular case, there was a variable named `rsaSignature` that but was an EdDSA signature.
+The fix is obvious: rename the variable to `eddsaSignature`.
```
Comment (mraible) [inline _source/_posts/2020-06-16-security-audits.adoc:43]: Obvious? Why not `edDsaSignature`? :trollface:

File: _source/_posts/2020-06-16-security-audits.adoc [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+Overall my experience with Paragon Initiative was great. They even found a typo in the project's README.md. A doc issue doesn't sound like much, but it shows their attention to detail while going through a codebase.
```
Comment (mraible) [inline _source/_posts/2020-06-16-security-audits.adoc:45]:
```suggestion
Overall my experience with Paragon Initiative was great. They even found a typo in the project's `README.md`. A doc issue doesn't sound like much, but it shows their attention to detail while going through a codebase.
```

File: _source/_posts/2020-06-16-security-audits.adoc [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+----
+public interface BasicMath {
+    int add(int a, int b);
+    int subtract(int b, int a);
+    int multiply(int a, int b);
+    int divide(int a, int b);
+}
+----
+
+In Java, the variable order maters. It's important to make sure all your methods use the same order. In the above example, the `subtract` methods arguments are `b, a` all of the other methods are ordered `a, b`. In isolation, this doesn' really matter, and the implementation might look like this:
```
Comment (mraible) [inline _source/_posts/2020-06-16-security-audits.adoc:63]:
```suggestion
In Java, the variable order matters. It's important to make sure all your methods use the same order. In the above example, the `subtract` methods arguments are `b, a` all of the other methods are ordered `a, b`. In isolation, this doesn't really matter, and the implementation might look like this:
```

File: _source/_posts/2020-06-16-security-audits.adoc [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+[source,java]
+----
+public class ExampleBasicMath {
+public int add(int a, int b) {
+return a + b;
+}
```
Comment (mraible) [inline _source/_posts/2020-06-16-security-audits.adoc:70]: This code block looks like it needs to be formatted.

File: _source/_posts/2020-06-16-security-audits.adoc (hunk 2 above)
Comment (bdemers) [reply _source/_posts/2020-06-16-security-audits.adoc:43]: good point!
do you think "obvious" is too strong? Maybe something like:
```suggestion
The fix is simple: rename the variable to `edDsaSignature `.
```

File: _source/_posts/2020-06-16-security-audits.adoc (hunk 1 above)
Comment (bdemers) [reply _source/_posts/2020-06-16-security-audits.adoc:38]: Nice! I didn't know that!

File: _source/_posts/2020-06-16-security-audits.adoc (hunk 2 above)
Comment (mraible) [reply _source/_posts/2020-06-16-security-audits.adoc:43]: Yeah, I think it's a bit strong.

**Comment by bdemers** [comment]
added to sec site

**Pull request closed without merging**